[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89981
* #89980
* __->__ #89900
* #89899
* #89898

**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern.

fegin committed:
[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"




[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"


**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern. 

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"


**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern. 

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"


**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern. 

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1323,3 +1372,100 @@ def return_fn(fqn_to_param_info):
+        value = orig_state[state_name]
+        if not isinstance(value, list) or not torch.is_tensor(value[0]):
+            continue
+        value = torch.concat(value)[: flat_param._numels[param_idx]].reshape(
```
Conceptual: In what case would `torch.concat(value)` have more numel than `flat_param._numels[param_idx]` (meaning that we are truncating)?

fegin replied:
Excellent point. The padding shouldn't be visible here. However, I'm not sure if it is always true as we were discussing the cost of F.pad and may change the implementation. This is just a safety guard. If you believe padding will never be seen, I can remove it.

awgu replied:
Ah good catch. Let us keep the trimming logic here. The safe guard is good.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1295,6 +1342,8 @@ def _optim_state_dict(
 def _get_fqn_to_fsdp_param_info(
     model: nn.Module, dedup_shared_fqns: Set[str]
 ) -> Dict[str, FSDPParamInfo]:
+    """Construct the mapping from fqn to the corrsponding FSDPParamInfo."""
```
nit:
```suggestion
    """Construct the mapping from fqn to the corresponding FSDPParamInfo."""
```

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1161,7 +1141,63 @@ def _map_param_id_to_optim_keys(
+    param_id_to_param: List[nn.Parameter],
+    param_to_fqns: Dict[nn.Parameter, List[str]],
+    fqn_to_fsdp_param_info: Dict[str, FSDPParamInfo],
+    merge_key: bool = False,
```
Do we need `merge_key=True` for `use_orig_params=True` to handle the fact that ranks who do not have any part of an original parameter will not have any optimizer state for that parameter?

Then, we also have `merge_key=False` to preserve the existing behavior for `use_orig_params=False`?

fegin replied:
That's correct.

awgu reviewed (commented):
Big thanks for working on this and getting through the crazy Cartesian product of code paths!

I made an initial pass and will continue to revisit.

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"


**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern. 

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"


**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern. 

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"


**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern. 

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"


**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern. 

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1086,50 +1100,16 @@ def _is_zero_dim_tensor(x: Any) -> bool:
-def _map_param_id_to_optim_keys(
-    optim_state_dict: Dict[str, Any],
-    group: Optional[dist.ProcessGroup],
+def _check_rank_missing_keys(
```
nit:
```suggestion
def _check_missing_keys_on_rank(
```

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1159,7 +1139,64 @@ def _map_param_id_to_optim_keys(
+    merge_keys: bool = False,
+) -> Tuple[List[_OptimStateKey], Dict[_OptimStateKey, int]]:
+    """
+    Construct the local mapping between the ``_OptimStateKey`` and parameter IDs
```
Thanks for adding this documentation!

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1238,34 +1276,44 @@ def _optim_state_dict(
-        param_id = optim_state_key_to_param_id[optim_state_key]
+    for optim_state_key in all_optim_state_keys:
+        param_id = optim_state_key_to_param_id.get(optim_state_key, -1)
+        assert param_id >= 0 or (optim_state_key.is_fsdp_managed and use_orig_params)
```
nit: Can we add a message to the `assert` explaining that for `use_orig_params=True`, some parameters may not be in the local shard, so they will be missing the optimizer state keys?

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1344,12 +1348,18 @@ def full_optim_state_dict(
             optim_input,
             optim,
         )
+        use_orig_params: bool = False
```
cc: @rohan-varma We were talking about which FSDP constructor args we assume to be uniform. This is one place evidencing that `use_orig_params` must be uniform (even if we already intuitively thought that already).

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1652,6 +1676,69 @@ def rekey_optim_state_dict(
+                were passed into the optimizer ``optim``.
+            optim (torch.optim.Optimizer): Optimizer for ``model`` 's
+                parameters.
+                the default process group. (Default: ``None``)
```
It looks like some part got cut off here for `group`?

awgu commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1652,6 +1676,69 @@ def rekey_optim_state_dict(
+                were passed into the optimizer ``optim``.
+            optim (torch.optim.Optimizer): Optimizer for ``model`` 's
+                parameters.
+            group (dist.ProcessGroup): Model's process group or ``None`` if using
```
nit: For both here and `_optim_state_dict`, let us mention that this is the process group across which parameters are sharded since we have landed HSDP, which has both the intra-node sharding process group and the inter-node replicating process group.
```suggestion
            group (dist.ProcessGroup): Model's process group across which parameters are sharded or ``None`` if using
```

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1329,3 +1377,100 @@ def return_fn(fqn_to_param_info):
+
+
+def _shard_orig_param_state(
+    fqn: str,
```
nit: For this function and `_gather_orig_param_state`, can we specify if `fqn` is the local FQN (as stored in `flat_param._fqns`) or the global FQN (which may require prepending a prefix starting from the local FSDP root).

Ugh, we might need to clarify our terminology at some point and just write it down somewhere.

awgu commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1329,3 +1377,100 @@ def return_fn(fqn_to_param_info):
+    object_list: List[Dict[str, Any]] = [
+        {} for _ in range(cast(int, fsdp_state.world_size))
+    ]
+    dist.all_gather_object(object_list, state_objects)
```
I forget; for the `use_orig_params=False` case, what collective do we use? `all_gather_into_tensor`?

fegin replied:
That's correct.

awgu reviewed (approved):
This looks good to me! Hopefully, you can get another review from someone else as well, but it is not strictly necessary.

I left some minor comments.

rohan-varma reviewed (approved):
LGTM, will let @awgu to accept it as well, thanks!

fegin commented on test/distributed/fsdp/test_fsdp_optim_state.py:
```diff
@@ -749,32 +750,44 @@ def test_scatter_full_optim_state_dict_transformer(self) -> None:
+        self._test_load_optim_state(
+            _ModelClass.NESTED,
+            use_multiple_param_groups=False,
+            halve_world_size=False,
```
More tests will be introduced after we make `FSDP.optim_state_dict` support all use cases.

fegin commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1159,7 +1139,64 @@ def _map_param_id_to_optim_keys(
+        is_fsdp_managed = isinstance(param, FlatParameter)
+        if is_fsdp_managed:
+            assert fqns[0] in fqn_to_fsdp_param_info
+        is_fsdp_managed = fqns[0] in fqn_to_fsdp_param_info
```
Yes, for `use_orig_params` case, this line is required.

fegin commented on torch/distributed/fsdp/_optim_utils.py:
```diff
@@ -1329,3 +1377,100 @@ def return_fn(fqn_to_param_info):
+                torch.cuda.device_count(),
+                cast(dist.ProcessGroup, fsdp_state.process_group),
+            )
+        value = value.cpu()
```
Yes, we do in the original code path. This is the inconsistency between `state_dict` and `optim_state_dict`. I had a PR to fix this. But it never land. Will introduce this later when I complete `optim_state_dict()`.

fegin commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -1652,6 +1676,69 @@ def rekey_optim_state_dict(
+        you know why this API exists and how this API works.
+
+        Returns the optimizer state. The state will be sharded or consolidated
+        based on ``state_dict_type`` set by :meth:`set_state_dict_type` or
```
We have not enforced this yet. But it will be enforced when `optim_state_dict` support all use cases.

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"


**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern. 

[ghstack-poisoned]

fegin commented:
@pytorchbot merge

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"


**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern. 

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"


**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern. 

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict][3/N] Support use_orig_param optim_state_dict (non-broadcast version)"


**What:**
This PR add the optim state_dict support of `use_orig_params` with rank0_only is False. rank0_only support will be added in a following PR. The design of this PR focus on the simplicity and may not have good performance, especially for optim state_dict loading. Since optim state_dict loading is only called once in the beginning of the training, performance is not the major concern. 

[ghstack-poisoned]

fegin commented:
@pytorchbot merge -f "The failing test is unrelated."
